url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/misc/NumberOperations.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2006-2009 Odysseus Software GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.odysseus.el.misc;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+import javax.el.ELException;
+
+/**
+ * Arithmetic Operations as specified in chapter 1.7.
+ * 
+ * @author Christoph Beck
+ */
+public class NumberOperations {
+	private final static Long LONG_ZERO = Long.valueOf(0L);
+
+	private final static boolean isDotEe(String value) {
+		int length = value.length();
+		for (int i = 0; i < length; i++) {
+			switch (value.charAt(i)) {
+				case '.':
+				case 'E':
+				case 'e': return true;
+			}
+		}
+		return false;
+	}
+
+	private final static boolean isDotEe(Object value) {
+		return value instanceof String && isDotEe((String)value);
+	}
+
+	private static final boolean isFloatOrDouble(Object value) {
+		return value instanceof Float || value instanceof Double;
+	}
+
+	private static final boolean isFloatOrDoubleOrDotEe(Object value) {
+		return isFloatOrDouble(value) || isDotEe(value);
+	}
+
+	private static final boolean isBigDecimalOrBigInteger(Object value) {
+		return value instanceof BigDecimal || value instanceof BigInteger;
+	}
+
+	private static final boolean isBigDecimalOrFloatOrDoubleOrDotEe(Object value) {
+		return value instanceof BigDecimal || isFloatOrDoubleOrDotEe(value);
+	}
+
+	public static final Number add(TypeConverter converter, Object o1, Object o2) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 70-75 into a separate method.